### ISSUE TITLE
Option «-r, --range <range>»

### ISSUE THREAD

**Issue by u20221022** [curl/curl#10334]
### I did this
1) `curl.exe -i -r -1 "https://www.roboform.com/dist/RoboForm-v9-Setup.exe"`
HTTP/1.1 206 Partial Content
Accept-Ranges: bytes
Cache-Control: private, max-age=0, must-revalidate
Content-Length: 2
Content-Range: bytes 0-1/30794728
Content-Type: application/octet-stream
ETag: "E8E3D50100000000909DBD63"
reason: OK
server: server
server-id: -

MZ

2) `curl.exe -i -r 30794727 "https://www.roboform.com/dist/RoboForm-v9-Setup.exe"`
Warning: A specified range MUST include at least one dash (-). Appending one
Warning: for you!
HTTP/1.1 206 Partial Content
Accept-Ranges: bytes
Cache-Control: private, max-age=0, must-revalidate
Content-Length: 1
Content-Range: bytes 30794727-30794727/30794728
Content-Type: application/octet-stream
ETag: "E8E3D50100000000909DBD63"
reason: OK
server: server
server-id: -

Warning: Binary output can mess up your terminal. Use "--output -" to tell
Warning: curl to output it to your terminal anyway, or consider "--output
Warning: <FILE>" to save to a file.
### I expected the following
The same thing, but without false threats as:
«Warning: A specified range MUST include at least one dash (-). Appending one
Warning: for you!»
and
«Warning: Binary output can mess up your terminal. Use "--output -" to tell
Warning: curl to output it to your terminal anyway, or consider "--output
Warning: <FILE>" to save to a file.»
### curl/libcurl version

[curl -V output]
curl 7.87.0 (x86_64-w64-mingw32) libcurl/7.87.0 OpenSSL/3.0.7 (Schannel) zlib/1.2.13 brotli/1.0.9 zstd/1.5.2 WinIDN libssh2/1.10.0 nghttp2/1.51.0 ngtcp2/0.12.1 nghttp3/0.8.0 libgsasl/2.2.0
Release-Date: 2022-12-21
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp ws wss
Features: alt-svc AsynchDNS brotli gsasl HSTS HTTP2 HTTP3 HTTPS-proxy IDN IPv6 Kerberos Largefile libz MultiSSL NTLM SPNEGO SSL SSPI threadsafe TLS-SRP UnixSockets zstd
### operating system
Windows 10 last version with all updates

**Comment by dfandrich** [comment]
RFC 2616 specifies that a range must contain a dash, so the message looks right to me.

**Comment by jay** [comment]
I agree, this is not a bug in curl.

I find it interesting that requesting a range of -1 (which should be last n bytes according to [suffix-byte-range-spec](https://www.rfc-editor.org/rfc/rfc2616#section-14.35.1)) causes the server to return the first two bytes instead of the last byte. The server seems to be treating it as byte-range-spec (0-1) instead of suffix-byte-range-spec (-1).

~~~
> GET /dist/RoboForm-v9-Setup.exe HTTP/1.1
> Host: www.roboform.com
> Range: bytes=-1
> User-Agent: curl/7.82.0-DEV
> Accept: */*
>
* STATE: DO => DID handle 0xd541d8; line 2145 (connection #0)
* STATE: DID => PERFORMING handle 0xd541d8; line 2264 (connection #0)
* Mark bundle as not supporting multiuse
* HTTP 1.1 or later with persistent connection
< HTTP/1.1 206 Partial Content
< Accept-Ranges: bytes
< Cache-Control: private, max-age=0, must-revalidate
< Content-Length: 2
< Content-Range: bytes 0-1/30794728
< Content-Type: application/octet-stream
< ETag: "E8E3D50100000000909DBD63"
< reason: OK
< server: server
< server-id: -
~~~

If anyone knows what "server: server" actually is we could report it.

**Comment by u20221022** [comment]
> If anyone knows what "server: server" actually is we could report it.

`curl.exe --http1.1 -I -L -w "%{JSON}" "https://www.rarlab.com/"`
HTTP/1.1 200
date: Mon, 23 Jan 2023 17:58:01 GMT
server: Apache
…

Non-standard server response – without «OK» after «200».  Report it.

**Comment by dfandrich** [comment]
Any text after the status code is optional and servers don't have to send anything. curl just reports what was sent.

**Comment by u20221022** [comment]
> Any text after the status code is optional

Where is the link to the proof?

> curl just reports what was sent

cURL has terribly many bugs.  That is the proof (for example):
1) `curl.exe --http3 --http2 --http1.1 --http1.0 --http0.9 -I -L -w "%{JSON}" "https://curl.se/windows/dl-7.87.0_2/curl-7.87.0_2-win64-mingw.zip"`
HTTP/1.1 200 OK
Connection: close
Content-Length: 10582157
Server: nginx/1.21.1
Content-Type: application/zip
X-Frame-Options: SAMEORIGIN
Last-Modified: Wed, 21 Dec 2022 06:05:38 GMT
ETag: "a1788d-5f05057677080"
Cache-Control: max-age=31536000
Expires: Wed, 10 Jan 2024 16:15:24 GMT
X-Content-Type-Options: nosniff
Content-Security-Policy: default-src 'self' curl.haxx.se www.curl.se curl.se www.fastly-insights.com fastly-insights.com; style-src 'unsafe-inline' 'self' curl.haxx.se www.curl.se curl.se
Strict-Transport-Security: max-age=31536000
Via: 1.1 varnish, 1.1 varnish
Accept-Ranges: bytes
Age: 1290938
Date: Wed, 25 Jan 2023 14:51:01 GMT
X-Served-By: cache-bma1661-BMA, cache-fra-eddf8230048-FRA
X-Cache: HIT, HIT
X-Cache-Hits: 26, 0
X-Timer: S1674658262.615398,VS0,VE1
alt-svc: h3=":443";ma=86400,h3-29=":443";ma=86400,h3-27=":443";ma=86400

{"content_type":"application/zip","errormsg":null,"exitcode":0,"filename_effective":null,"ftp_entry_path":null,"http_code":200,"http_connect":0,"http_version":"1.1","local_ip":"192.168.1.34","local_port":39534,"method":"HEAD","num_connects":1,"num_headers":21,"num_redirects":0,"proxy_ssl_verify_result":0,"redirect_url":null,"referer":null,"remote_ip":"151.101.65.91","remote_port":443,"response_code":200,"scheme":"HTTPS","size_download":0,"size_header":875,"size_request":121,"size_upload":0,"speed_download":0,"speed_upload":0,"ssl_verify_result":0,"time_appconnect":1.630744,"time_connect":0.237204,"time_namelookup":0.023485,"time_pretransfer":1.393847,"time_redirect":0.000000,"time_starttransfer":1.524329,"time_total":1.643068,"url":"https://curl.se/windows/dl-7.87.0_2/curl-7.87.0_2-win64-mingw.zip","url_effective":"https://curl.se/windows/dl-7.87.0_2/curl-7.87.0_2-win64-mingw.zip","urlnum":0,"curl_version":"libcurl/7.87.0 OpenSSL/3.0.7 (Schannel) zlib/1.2.13 brotli/1.0.9 zstd/1.5.2 WinIDN libssh2/1.10.0 nghttp2/1.51.0 ngtcp2/0.12.1 nghttp3/0.8.0 libgsasl/2.2.0"}

2) `curl.exe --http0.9 --http1.0 --http1.1 --http2 --http3 -I -L -w "%{JSON}" "https://curl.se/windows/dl-7.87.0_2/curl-7.87.0_2-win64-mingw.zip"`
HTTP/3 200
content-length: 10582157
server: nginx/1.21.1
content-type: application/zip
x-frame-options: SAMEORIGIN
last-modified: Wed, 21 Dec 2022 06:05:38 GMT
etag: "a1788d-5f05057677080"
cache-control: max-age=31536000
expires: Wed, 10 Jan 2024 16:15:24 GMT
x-content-type-options: nosniff
content-security-policy: default-src 'self' curl.haxx.se www.curl.se curl.se www.fastly-insights.com fastly-insights.com; style-src 'unsafe-inline' 'self' curl.haxx.se www.curl.se curl.se
strict-transport-security: max-age=31536000
via: 1.1 varnish, 1.1 varnish
accept-ranges: bytes
age: 1290973
date: Wed, 25 Jan 2023 14:51:36 GMT
x-served-by: cache-bma1661-BMA, cache-fra-eddf8230109-FRA
x-cache: HIT, HIT
x-cache-hits: 26, 0
x-timer: S1674658297.864228,VS0,VE1
alt-svc: h3=":443";ma=86400,h3-29=":443";ma=86400,h3-27=":443";ma=86400

{"content_type":"application/zip","errormsg":null,"exitcode":0,"filename_effective":null,"ftp_entry_path":null,"http_code":200,"http_connect":0,"http_version":"3","local_ip":"192.168.1.34","local_port":53819,"method":"HEAD","num_connects":2,"num_headers":20,"num_redirects":0,"proxy_ssl_verify_result":0,"redirect_url":null,"referer":null,"remote_ip":"151.101.65.91","remote_port":443,"response_code":200,"scheme":"HTTPS","size_download":0,"size_header":852,"size_request":119,"size_upload":0,"speed_download":0,"speed_upload":0,"ssl_verify_result":0,"time_appconnect":0.000000,"time_connect":1.220702,"time_namelookup":0.021376,"time_pretransfer":1.220937,"time_redirect":0.000000,"time_starttransfer":2.221241,"time_total":2.365529,"url":"https://curl.se/windows/dl-7.87.0_2/curl-7.87.0_2-win64-mingw.zip","url_effective":"https://curl.se/windows/dl-7.87.0_2/curl-7.87.0_2-win64-mingw.zip","urlnum":0,"curl_version":"libcurl/7.87.0 OpenSSL/3.0.7 (Schannel) zlib/1.2.13 brotli/1.0.9 zstd/1.5.2 WinIDN libssh2/1.10.0 nghttp2/1.51.0 ngtcp2/0.12.1 nghttp3/0.8.0 libgsasl/2.2.0"}

All «--http*» options are «mutually exclusive».  Where is the error #2 or #3?  The manual was completely rewritten a few days ago.

**Comment by dfandrich** [comment]
> Where is the link to the proof?

You're the one alleging that curl is doing something wrong. The burden of proof is on you.

> cURL has terribly many bugs. That is the proof (for example):

You'll have to explain where the bug is. Dumping dozens of lines of headers
without explanation doesn't help.

> All «--http*» options are «mutually exclusive». Where is the error #2 or #3?

The manual explains how the options work. I don't know what error you're
talking about.

> The manual was completely rewritten a few days ago.

No it wasn't.

**Comment by u20221022** [comment]
Читать научись, а потом умничай.  Между `curl.exe --manual` версий «7.86.0» и «7.87.0» более 150 различий.  Вот первые из них (по одной строке, без переносов).
1) -o, --output or -O, --remote-name options. If curl  is  given  multiple
--output or --remote-name options. If curl is given  multiple  URLs  to
2) left, etc. The progress meter displays the transfer rate in  bytes  per
left, etc. The progress meter displays number of bytes and  the  speeds
3) redirect the response output to a file, using shell redirect  (>),  -o,
redirect the response output to  a  file,  using  shell  redirect  (>),
4) If you prefer a progress  "bar"  instead  of  the  regular  meter,  -#,
If  you  prefer  a  progress  "bar"  instead  of  the  regular   meter,
5) is a recommended separator. The long "double-dash" form, -d, --data for
is a recommended separator. The long "double-dash" form, --data for ex-

Хватит меня спамить, Дфандрич.

**Comment by u20221022** [comment]
> The manual was completely rewritten a few days ago.
> No it wasn't.

«Bulk», NOT «completely».  Can you read at all?

**Comment by dfandrich** [comment]
First, 7.87.0 was released 35 days ago and 7.86.0 92 days ago. I wouldn't call that "a few days ago". Second, 2.5% of words in `curl --manual` were changed between those two versions. I wouldn't call that "completely rewritten" or even "bulk rewritten" (and even if you edited your comment, you did originally write "completely"). Third, if you didn't want a response to this issue, why did you open it?

**Comment by u20221022** [comment]
35 дней – это несколько лет?  Или несколько месяцев?  Переделка была вызвана моим сообщением о недостатках документации.  Все изменения были сделаны за одну-две недели.  Полторы сотни исправленных мест за такое время – это по десятку в день.  Ты хотел бы, чтобы каждый день переделывали половину документации?
Я воспользовался компьютерным переводчиком, который в мою фразу вставил слово «completely».  Я заметил и исправил эту ошибку.  Я что, должен был полностью удалять комментарий?  Для чего на этом сайте существует возможность исправления?  Читай свою почту внимательнее, раз уж браузером – который показывает последнюю версию – пользоваться не научился.
И что ты врёшь, будто бы я приписываю ошибку сервера программе cURL?  Вот твои слова: «You're the one alleging that curl is doing something wrong. The burden of proof is on you.»  Я писал об ошибках сервера.  Сервера, а не cURL.  Если ты знаешь, где конкретно указано, что расшифровка кода ответа сервера не является обязательной даже в протоколе HTTP/1.1, почему ты не указываешь это прямо?  Я в два раза старше тебя и в миллион раз умнее.  Отстань от меня.
